kernel/kernel-api/src/main/java/io/delta/kernel/SnapshotBuilder.java

```diff
@@ -117,6 +117,38 @@ public interface SnapshotBuilder {
   // TODO: [delta-io/delta#4821] Public Metadata API
   SnapshotBuilder withProtocolAndMetadata(Protocol protocol, Metadata metadata);
 
+  /**
+   * Specifies the maximum table version known by the catalog.
+   *
+   * <p>This method is used by catalog implementations for catalog-managed Delta tables to indicate
+   * the highest version of the table that the catalog is aware of. This ensures that any snapshot
+   * resolution operations respect the catalog's view of the table state.
+   *
+   * <p><b>Important:</b> This method is <b>required</b> for catalog-managed tables and <b>must not
+   * be used</b> for file-system managed tables. An {@link IllegalArgumentException} will be thrown
+   * at build time if this constraint is violated.
+   *
+   * <p>When specified, the following additional constraints are enforced:
+   *
+   * <ul>
+   *   <li>If {@link #atVersion(long)} is used for time travel, the requested version must be less
+   *       than or equal to the max catalog version.
+   *   <li>If {@link #atTimestamp(long, Snapshot)} is used for time travel, the provided {@code
+   *       latestSnapshot} must have a version equal to the max catalog version.
+   *   <li>If {@link #withLogData(List)} is provided and {@link #atVersion(long)} is used, the log
+   *       data must include the requested version (i.e., the tail of the log data must have a
+   *       version greater than or equal to the requested version).
+   *   <li>If {@link #withLogData(List)} is provided and no version is specified (resolving to
+   *       latest), the log data must end with the max catalog version.
+   * </ul>
+   *
+   * @param version the maximum table version known by the catalog (must be >= 0)
+   * @return a new builder instance with the specified max catalog version
+   * @throws IllegalArgumentException if version is negative
+   */
+  // TODO revisit specific naming
```

Comment:
@zachschuermann I think you had some ideas on this?

Reply:
I think it is ideal to include "catalog" of some sort considering this _should not be used_ for filesystem tables. Also, we use `with` for all our builder patterns so I don't think it makes sense to skip that.

I'm indifferent on max vs latest as long as we choose and standardize on one.

cc @scottsand-db @nicklan

Reply:
+1 on having `catalog` in the name.